backend/logic/courses/courses.js

```diff
@@ -1,19 +1,153 @@
+  }
+};
+
+/**
+ * Adds a student's completed courses into the database
+ * @author Steven Li
+ * @param {int} studentId 
+ * @param {JSON} courses 
+ * @returns true if successful
+ * @throws error if MySQL connection failed
+ *         invalid format courses if JSON format is incorrect
+ *         false if insertion failed
+ */
+var addCompletedCourses = async (studentId, courses) => {
+  await format.verifyCourse(courses);
+  let connection = await mysql.getNewConnection();
+
+  try {
+    // Obtain all current course offerings and put them into a hashtable for fast lookup
+    let results = await connection.query(
+      "SELECT course_code, semester, section, id FROM course_offerings;"
+    );
+    let hashTable = {};
+    for (let i = 0; i < results.length; i++) {
+      hashTable[
+        results[i].course_code + results[i].semester + results[i].section
+      ] = results[i].id;
+    }
+
+    await connection.beginTransaction();
+    // Insert each course of the student into the database. If a course does not exist then
+    // throw an error.
+    for (let course in courses) {
+      for (let i = 0; i < courses[course].length; i++) {
+        let id =
+          hashTable[
+            course + courses[course][i].semester + courses[course][i].section
+          ];
+        await connection.query(
+          "INSERT INTO student_course_offerings VALUES(?, ?, ?);",
+          [studentId, id, courses[course][i].semester]
+        );
+      }
     }
 
-    let connection = await mysql.getNewConnection();
-    try {
-        let courses = await connection.query('SELECT course_code FROM course_tag WHERE course_tag.tag_name LIKE ?', tag);
-        connection.release();
-        return (JSON.parse(JSON.stringify(courses)));
-    } catch (error) {
-        connection.release();
-        console.error(error);
-        throw Error(error.message);
+    await connection.commit();
+    connection.release();
+    return true;
+  } catch (error) {
+    console.error(error);
+    await connection.rollback();
+    connection.release();
+    throw new Error(false);
+  }
+};
+
+/**
+ * Returns all available courses from the database
+ * @author Steven Li
+ * @returns a list of available courses from the database in JSON format
+ * @throws error if MySQL connection failed
+ */
+var getAllCourses = async () => {
+  let connection = await mysql.getNewConnection();
+  let courses;
+  let result;
+  try {
+    result = await connection.query("SELECT * FROM courses;");
+    connection.release();
+  } catch (err) {
+    console.error(err);
+    connection.release();
+  }
+
+  if (result) {
+    courses = result;
+  }
+  return courses;
+};
+
+/**
+ * Adds a list of course offerings into the database
+ * @author Steven Li
+ * @param {JSON} courseOfferings
+ * @returns true if successful
+ * @throws error if MySQL connection failed
+ *         invalid format course offerings if JSON format is wrong
+ *         false if insertion failed
+ */
```

Comment:
Should we also add the course associated with the course offering if it doesnt exit? Or have the user do that himself with the other endpoint